types/underscore/index.d.ts

```diff
@@ -451,70 +451,42 @@ declare module _ {
 
         /**
-        * Returns the maximum value in list.
-        * @param list Finds the maximum value in this list.
-        * @return Maximum value in `list`.
-        **/
-        max(list: _.List<number>): number;
-
-        /**
-        * @see _.max
-        */
-        max(object: _.Dictionary<number>): number;
-
-        /**
-        * Returns the maximum value in list. If iterator is passed, it will be used on each value to generate
-        * the criterion by which the value is ranked.
-        * @param list Finds the maximum value in this list.
-        * @param iterator Compares each element in `list` to find the maximum value.
-        * @param context `this` object in `iterator`, optional.
-        * @return The maximum element within `list`.
-        **/
-        max<T>(
-            list: _.List<T>,
-            iterator?: _.ListIterator<T, any>,
-            context?: any): T;
-
-        /**
-        * @see _.max
-        */
-        max<T>(
-            list: _.Dictionary<T>,
-            iterator?: _.ObjectIterator<T, any>,
-            context?: any): T;
-
-        /**
-        * Returns the minimum value in list.
-        * @param list Finds the minimum value in this list.
-        * @return Minimum value in `list`.
-        **/
-        min(list: _.List<number>): number;
-
-        /**
-         * @see _.min
-         */
-        min(o: _.Dictionary<number>): number;
-
-        /**
-        * Returns the minimum value in list. If iterator is passed, it will be used on each value to generate
-        * the criterion by which the value is ranked.
-        * @param list Finds the minimum value in this list.
-        * @param iterator Compares each element in `list` to find the minimum value.
-        * @param context `this` object in `iterator`, optional.
-        * @return The minimum element within `list`.
-        **/
-        min<T>(
-            list: _.List<T>,
-            iterator?: _.ListIterator<T, any>,
-            context?: any): T;
+         * Returns the maximum value in `collection`. If an `iteratee` is
+         * provided, it will be used on each element to generate the criterion
+         * by which the element is ranked. -Infinity is returned if list is
+         * empty. Non-numerical values returned by `iteratee` will be ignored.
+         * @param collection The collection in which to find the maximum value.
+         * @param iteratee The iteratee that provides the criterion by which
+         * each element is ranked, optional if evaluating a collection of
+         * numbers.
+         * @param context `this` object in `iteratee`, optional.
+         * @returns The maximum element within `collection` or -Infinity if
+         * `collection` is empty.
+         **/
+        max<V extends Collection<any>>(
+            collection: V,
+            iteratee?: Iteratee<V, number>,
+            context?: any
+        ): TypeOfCollection<V> | number;
```

Comment:
I think, if you want to be very precise, you can define some helper types,

```ts
interface NumberComparable {
    valueOf(): number;
}

type ValueOrLimit<I, T, L> = Extract<
    IterateeResult<I, T>,
    NumberComparable
> extends never ? L : T | L;
```

and then use the following signature:

```suggestion
        max<V extends Collection<any>, I extends Iteratee<V, number> = undefined>(
            collection: V,
            iteratee?: I,
            context?: any
        ): ValueOrLimit<I, TypeOfCollection<V>, -Infinity>;
```

This is assuming that `number extends NumberComparable`. Otherwise, you need to use `number | NumberComparable` as the second parameter to `Extract`.

Reply:
Huh, interesting, similar to [toPrimitive](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/46068#discussion_r454887492) I also didn't realize that `valueOf` was a thing. Happily `number extends NumberComparable` is true, so I'll go with that and make these changes, thanks!

Reply:
Hmm, in trying this out though it doesn't work with `boolean`s even though `_.max([true, false])` ends up providing `true`, and while `_.max(['a', 'b', 'c'])` results in `-Infinity`, `_.max(['1', '2', '3'])` results in `3`. Something else to consider is that the actual objects passed to `max` might implement `NumberComparable` even if their declared type ends up being a subset of the object's properties that does not implement `NumberComparable`.

While these types might possibly be a little more descriptive I think they end up being less accurate due to some of these edge cases, and even the increase in descriptiveness is debatable since the types involved are not straightforward. Overall I think I'd prefer to leave this as-is at this point unless you strongly object, and I'm admittedly even a little tempted to change `Iteratee<V, number>` to `Iteratee<V, any>` to allow for more of these edge cases as iteratee results (which is something that at least some of the previous overloads allowed for).

Reply:
On a related note, I've also been waffling a little for a little while now about it not being entirely correct to limit boolean-result iteratees like `find` and `some` to only boolean return types, and this discussion is making me lean even more toward removing restrictions on iteratee return types across the board. The reasons I'm not entirely happy with those limitations are twofold. First, the restriction currently isn't being consistently applied; it really only applies to function-type iteratees specifically. Second, it's not important to Underscore that they be converted to booleans; Underscore will evaluate the provided values on their truthiness and falsiness.

Reply:
Another edge case worth noting here is the use of a `toPrimitive` overload rather than a `valueOf` overload since `_.max([{ [Symbol.toPrimitive]: () => 1, a: 'a' }, { [Symbol.toPrimitive]: () => 2, a: 'b' }, { [Symbol.toPrimitive]: () => 3, a: 'c' }])` results in `{a: "c", Symbol(Symbol.toPrimitive): ƒ}`.

Reply:
Per my comments above @jgonggrijp I've updated the expected return type for `max` and `min` iteratees from `number` to `any`; let me know if you disagree with that change.